aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -654,6 +654,21 @@ TensorIterator TensorIterator::binary_op(Tensor& out, const Tensor& a,
   return iter;
 }
 
+TensorIterator TensorIterator::comparison_op(Tensor& out, const Tensor& a,
+    const Tensor& b, bool check_mem_overlap) {
+  auto iter = TensorIterator();
+  iter.set_check_mem_overlap(check_mem_overlap);
+  iter.add_output(out);
+  iter.add_input(a);
+  iter.add_input(b);
+  iter.allow_cpu_scalars_ = true;
+  if (!out.is_same(a) && !out.is_same(b)) {
```

Comment:
Is this correct? Let's say we're doing
```
x = torch.tensor(..., dtype=int32)
y = torch.tensor(..., dtype=int64)
torch.lt(x, y, out=x)
```
In `lt_kernel`, `return_t` is int32. In addition, `scalar_t` is `iter.input_dtype()`, which is equal to the dtype of the first input. The first input is `x`, so `scalar_t` is int32. However, we should really be comparing `x` and `y` in int64.

We should add a test case for this situation by creating two numbers `x: int32` and `y: int64` such that their comparison returns a different result depending on if they're being compared in int32 or int64.

Reply:
From the test it looks like the behavior is correct. However, I still think this code snippet is confusing. How does tensor iterator know that `input_dtype` is torch.int32 in the following example?
```
x = torch.tensor([1 << 5], dtype=torch.uint8, device=device)
y = torch.tensor([1 << 15], dtype=torch.int32, device=device)
result = torch.lt(x, y, out=x)
self.assertEqual(result, torch.tensor([True]))
```

Reply:
Type will be promoted based on: https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h#L386 to the 'highest' type. So uint8 and int32 will be promoted to int32 and then will be compared and set to out

Reply:
If that is the logic, do we even need to call `iter.compute_common_dtype_only_for_inputs();`?

Reply:
Let me rephrase that. It seems strange to me that if any of the outputs are the same as the input, we use CommonDTypeStrategy::COMPUTE_ALL, but if the outputs are distinct from the inputs, we use CommonDTypeStrategy::COMPUTE_INPUTS. Knowing nothing about the implementation, it feels like we should unconditionally use CommonDTypeStrategy::COMPUTE_INPUTS for comparison operators because comparison operators should not use their outputs to inform decision about type promotion.

Reply:
Currently in TensorIterator with COMPUTE_ALL strategy we prevent to automatically cast from 'bigger' to 'smaller' type, so TensorIterator will throw an error if, for example, we will try to cast from Float to Bool. But in case of logical operation we can do this and to avoid this restriction I promote input types only then cast result to Bool in lambda and populate to output.

Reply:
I made changes based on our offline discussion

Reply:
Resolving this, thanks for making the change.